x-pack/plugins/alerting/server/lib/determine_flapping.ts

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { keys } from 'lodash';
+import { Alert } from '../alert';
+import { AlertInstanceState, AlertInstanceContext, RawAlertInstance } from '../types';
+
+const MAX_CAPACITY = 20;
+const MAX_FLAP_COUNT = 4;
+
+export function determineFlapping<
+  State extends AlertInstanceState,
+  Context extends AlertInstanceContext,
+  ActionGroupIds extends string
+>(activeAlerts: Record<string, Alert<State, Context, ActionGroupIds>> = {}): Set<string> {
+  const flappingAlertIds = new Set<string>();
+
+  for (const id of keys(activeAlerts)) {
+    const alert = activeAlerts[id];
+    if (isFlapping(alert)) {
+      flappingAlertIds.add(id);
+    }
+  }
+
+  return flappingAlertIds;
+}
+
+export function determineAlertsToReturn<
+  State extends AlertInstanceState,
+  Context extends AlertInstanceContext,
+  ActionGroupIds extends string,
+  RecoveryActionGroupId extends string
+>(
+  activeAlerts: Record<string, Alert<State, Context, ActionGroupIds>> = {},
+  recoveredAlerts: Record<string, Alert<State, Context, RecoveryActionGroupId>> = {}
+): {
+  alertsToReturn: Record<string, RawAlertInstance>;
+  recoveredAlertsToReturn: Record<string, RawAlertInstance>;
+} {
+  const alertsToReturn: Record<string, RawAlertInstance> = {};
+  const recoveredAlertsToReturn: Record<string, RawAlertInstance> = {};
+
+  for (const id of keys(activeAlerts)) {
+    alertsToReturn[id] = activeAlerts[id].toRaw();
+  }
+
+  for (const id of keys(recoveredAlerts)) {
+    const alert = recoveredAlerts[id];
+    // return recovered alerts if they are flapping or if the flapping array is not at capacity
+    if (isFlapping(alert)) {
+      recoveredAlertsToReturn[id] = alert.toRaw(true);
+    } else if (!atCapacity(alert.getFlappingHistory())) {
```

Comment:
This was something I did to save space in the state. I would stop track a recovered alert if it wasn't flapping in the last 20 executions. So the next time it was active it would be treated like a new alert

Reply:
yes, i think so too! I will also add a comment here

Reply:
Resolved in this commit https://github.com/elastic/kibana/pull/144415/commits/a02d7bb325820ab38295e0b94225c1ec00124e66